user_guide_src/source/database/results.rst

```diff
@@ -166,6 +166,31 @@ it returns the current row and moves the internal data pointer ahead.
         echo $row->body;
     }
 
+For use with MySQLi you must set MySQLi's result mode to 
+``MYSQLI_USE_RESULT``.
```

Comment:
According to my tests the method does not work properly without using MYSQLI_USE_RESULT. It isn't truly an unbuffered query unless you use it. I don't quite understand why this was implemented without it. I ran a query that consumed 100+mb using the current method and then after using MYSQLI_USE_RESULT it was down to 5mb. So I'm not sure what using getUnbufferedRow() is useful for without also using MYSQLI_USE_RESULT.

Reply:
It is working correctly, but maybe not how you would expect it to work. As I mentioned earlier - even without changing the result mode this method will give developers some advantages. It's about not returning the whole result at once.

The initial implementation was meant to make it possible to iterate over huge results on the PHP side, not to optimize the query on the database side.

Reply:
The only thing I can see is that using getUnbuggeredRow() as it is will only put one row at a time into a variable. But it should be noted that the entire result set is already in php memory.

From php docs:

MYSQLI_STORE_RESULT (default) - returns a mysqli_result object with buffered result set.

MYSQLI_USE_RESULT - returns a mysqli_result object with unbuffered result set.

As I understand this is that when query is called with the default setting the entire result set is retrieved by php from the MySQL server. This allows functions like seek, row count, etc. Php can scan through the entire result set and any given row can be fetched to a variable. When calling with Use Result, the communication channel remains open and php can request a row at a time until all rows are consumed and the channel closes. Under these circumstances php is unable to get row count and seek though the data. 

So this doesn't have anything to do with optimizing the database side. In fact its less than optimal on the database side because the database retains a read on the tables.  You said "It's about not returning the whole result at once." but it clearly does return the results to php at once even if you don't fetch them all into variables.

I'm just trying to get on the same page with you here.. I'm not sure how to document the advantages of using getUnbufferedRow() without also using MYSQLI_USE_RESULT.  I guess there is a slight advantage in that your not fetching the result as variables.

Reply:
> So this doesn't have anything to do with optimizing the database side. In fact its less than optimal on the database side because the database retains a read on the tables. 

We were talking about optimizing memory usage, so from this perspective buffer uses less memory - this is what I was referring to.

> You said "It's about not returning the whole result at once." but it clearly does return the results to php at once even if you don't fetch them all into variables.

Not exactly, it returns the whole result to the buffer and that's a big difference - in the memory usage. We may not understand each other correctly here. One thing is to return the entire result (or not) to the buffer, and the other thing is to return the buffer to PHP variables.

> I'm just trying to get on the same page with you here.. I'm not sure how to document the advantages of using getUnbufferedRow() without also using MYSQLI_USE_RESULT. I guess there is a slight advantage in that your not fetching the result as variables.

You've already done this by adding a notes section to your user guide. I think that's enough.

---

Overall, I have no problems with your arguments - you're right. What I'm trying to do here is to ensure that developers will change the result mode only if they know what they're doing. That's all.

When we use buffered mode the whole result is kept in the buffer and usually, it's not a problem. Fetching data from the buffer into PHP is the moment when problems occur.

To summarize this, when we operate on large results, we can describe methods this way:

* `$query->getResult()` - the most memory consuming and may lead to errors with exhausted allowed memory size
* `$query->getUnbufferedRow()` - safe choice, you will be able to iterate over your results without sacrificing any functionality
* `$query->getUnbufferedRow() with MYSQLI_USE_RESULT` - best choice if you're concerned about memory usage, but it comes with some trade-offs

Reply:
Thanks, I think I'm understanding now. I had originally opened an issue about this as a bug. I was really thrown off by the method name. Looking at how its implemented and reading through the guide I kind of see how it all goes together.  I have an idea of how to reword things a bit so it doesn't interfere with the current convention.  I'm thinking of presenting the use case of writing a query out to csv. This is how I came about the problem myself. I just couldn't understand why a buffered query would consume so much memory.  This feature definitely is very useful.. but yeah you wouldn't want to use it in most circumstances.  
I'll put something together for the userguide tomorrow.